src/box/xrow_io.cc

```diff
@@ -102,3 +105,54 @@ coio_write_xrow(struct iostream *io, const struct xrow_header *row)
 		diag_raise();
 }
 
+uint64_t xrow_stream_flush_size = 16384;
+TWEAK_UINT(xrow_stream_flush_size);
+
+void
+xrow_stream_write(struct xrow_stream *stream, const struct xrow_header *row)
+{
+	/* Fixheader - length of packet. */
+	size_t fixheader_len = mp_sizeof_uint(UINT32_MAX);
+	size_t len = 0;
+	len += xrow_header_sizeof(row, row->sync);
+	for (int i = 0; i < row->bodycnt; i++) {
+		len += row->body[i].iov_len;
+	}
+	char *data = (char *)xlsregion_alloc(&stream->lsregion,
+					     len + fixheader_len,
+					     ++stream->lsr_id);
+	*(data) = 0xce; /* MP_UINT32 */
+	store_u32(data + 1, mp_bswap_u32(len));
+	data = data + fixheader_len;
+	data += xrow_header_encode(row, row->sync, data);
+	for (int i = 0; i < row->bodycnt; i++) {
+		size_t l = row->body[i].iov_len;
+		memcpy(data, row->body[i].iov_base, l);
+		data += l;
+	}
+}
+
+int
+xrow_stream_flush(struct xrow_stream *stream, struct iostream *io)
+{
+	ssize_t to_flush = lsregion_used(&stream->lsregion);
+	/*
+	 * Might flush more than requested if data is added to the buffer
+	 * during the coio_writev yield.
+	 */
+	while (to_flush > 0) {
+		struct iovec iov[IOV_MAX];
+		int iovcnt = lengthof(iov);
+		int64_t gc_id = lsregion_to_iovec(&stream->lsregion, iov,
+						  &iovcnt, &stream->flush_pos);
+		ssize_t written = coio_writev(io, iov, iovcnt, 0);
+		if (written < 0)
+			return -1;
+		to_flush -= written;
+		lsregion_gc(&stream->lsregion, gc_id);
```

Comment:
I know, not possible, at least yet, but imagine that 2 fibers enter this place. One remembers `to_flush`, second remembers `to_flush + offset` where `offset` is how much the second fiber added. Now they both will spin in `while(to_flush > 0)` trying to send first `to_flush` bytes twice, no? The second one will be stuck in the loop, because combined they want to send `to_flush * 2 + offset`, but after first fiber finishes sending, the lsregion only has left `offset` bytes.

Reply:
Yes, indeed, it's possible. But haven't we agreed that only a single fiber can perform the flush? Won't it be the case after the relay rework?

It seems this is not the only problem here if we want to allow multiple fibers inside `xrow_stream_flush`. Even if I track `to_flush` correctly, the following sequence of events is possible:
1. fiber1 enters `xrow_stream_flush()`, sees lsregion contents `"abc"`, and remembers `to_flush = 3`.
2. fiber1 performs `lsregion_to_iovec` + `coio_writev()`, which yields after writing `"ab"`. `lsregion_svp` also has `pos = 3`.
3. Someone adds `"d"` to the stream.
4. fiber2 enters `xrow_stream_flush()`, sees lsregion contents "abcd" and `lsregion_svp` `pos = 3`, so it remembers `to_flush = 1`.
5. fiber2 performs `lsregion_to_iovec + `coio_writev()`, which writes `"d"` successfully. So as a result the receiver sees `"abdc"` instead of `"abcd"`.

Is this possible?

I could add guards against this, like that:

<details><summary>diff</summary>

```diff
diff --git a/src/box/xrow_io.cc b/src/box/xrow_io.cc
index 0dae40b60a..7f895786a2 100644
--- a/src/box/xrow_io.cc
+++ b/src/box/xrow_io.cc
@@ -36,6 +36,7 @@
 #include "errinj.h"
 #include "error.h"
 #include "msgpuck/msgpuck.h"
+#include "scoped_guard.h"
 #include "tweaks.h"
 
 void
@@ -135,6 +136,13 @@ xrow_stream_write(struct xrow_stream *stream, const struct xrow_header *row)
 int
 xrow_stream_flush(struct xrow_stream *stream, struct iostream *io)
 {
+#ifndef NDEBUG
+       assert(stream->owner == NULL);
+       stream->owner = fiber();
+       auto guard = make_scoped_guard([&] {
+               stream->owner = NULL;
+       });
+#endif
        ssize_t to_flush = lsregion_used(&stream->lsregion);
        /*
         * Might flush more than requested if data is added to the buffer
diff --git a/src/box/xrow_io.h b/src/box/xrow_io.h
index 48eb8b5bad..dfd31cbf06 100644
--- a/src/box/xrow_io.h
+++ b/src/box/xrow_io.h
@@ -68,6 +68,10 @@ struct xrow_stream {
        int64_t lsr_id;
        /** A savepoint used between flushes. */
        struct lsregion_svp flush_pos;
+#ifndef NDEBUG
+       /** A fiber which's currently using the stream. */
+       struct fiber *owner;
+#endif
 };
 
 /** Initialize the stream. */
@@ -82,6 +86,7 @@ xrow_stream_create(struct xrow_stream *stream)
 static inline void
 xrow_stream_destroy(struct xrow_stream *stream)
 {
+       assert(stream->owner == NULL);
        lsregion_destroy(&stream->lsregion);
 }
 
```

</details>

But I don't think it makes sense. We don't have such guards in coio or anywhere else. Haven't applied the change yet.

Reply:
I would add the guard. It wouldn't hurt and would prevent questions like I had above, because the guarantee would be clearly in the code and would be covered by tests.

Reply:
Ok, I've applied the diff above.
Resolving, since the PR is approved.